pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/JsonLogTransformer.java

```diff
@@ -0,0 +1,534 @@
+ * field.
+ * <p>
+ * For example, consider this log event:
+ * <pre>
+ * {
+ *   "timestamp": 1687786535928,
+ *   "hostname": "host1",
+ *   "level": "INFO",
+ *   "message": "Started processing job1",
+ *   "tags": {
+ *     "platform": "data",
+ *     "service": "serializer",
+ *     "params": {
+ *       "queueLength": 5,
+ *       "timeout": 299,
+ *       "userData_noIndex": {
+ *         "nth": 99
+ *       }
+ *     }
+ *   }
+ * }
+ * </pre>
+ * And let's say the table's schema contains these fields:
+ * <ul>
+ *   <li>timestamp</li>
+ *   <li>hostname</li>
+ *   <li>level</li>
+ *   <li>message</li>
+ *   <li>tags.platform</li>
+ *   <li>tags.service</li>
+ *   <li>indexableExtras</li>
+ *   <li>unindexableExtras</li>
+ * </ul>
+ * <p>
+ * Without this transformer, the entire "tags" field would be dropped when storing the record in the table. However,
+ * with this transformer, the record would be transformed into the following:
+ * <pre>
+ * {
+ *   "timestamp": 1687786535928,
+ *   "hostname": "host1",
+ *   "level": "INFO",
+ *   "message": "Started processing job1",
+ *   "tags.platform": "data",
+ *   "tags.service": "serializer",
+ *   "indexableExtras": {
+ *     "tags": {
+ *       "params": {
+ *         "queueLength": 5,
+ *         "timeout": 299
+ *       }
+ *     }
+ *   },
+ *   "unindexableExtras": {
+ *     "tags": {
+ *       "userData_noIndex": {
+ *         "nth": 99
+ *       }
+ *     }
+ *   }
+ * }
+ * </pre>
+ * Notice that the transformer:
+ * <ul>
+ *   <li>Flattens nested fields which exist in the schema, like "tags.platform"</li>
+ *   <li>Moves fields which don't exist in the schema into the "indexableExtras" field</li>
+ *   <li>Moves fields which don't exist in the schema and have the suffix "_noIndex" into the "unindexableExtras"
+ *   field</li>
+ * </ul>
+ * <p>
+ * The "unindexableExtras" field allows the transformer to separate fields which don't need indexing (because they are
+ * only retrieved, not searched) from those that do. The transformer also has other configuration options specified in
+ * {@link JsonLogTransformerConfig}.
+ * <p>
+ * One notable complication that this class handles is adding nested fields to the "extras" fields. E.g., consider
+ * this record
+ * <pre>
+ * {
+ *   a: {
+ *     b: {
+ *       c: 0,
+ *       d: 1
+ *     }
+ *   }
+ * }
+ * </pre>
+ * Assume "$.a.b.c" exists in the schema but "$.a.b.d" doesn't. This class processes the record recursively from the
+ * root node to the children, so it would only know that "$.a.b.d" doesn't exist when it gets to "d". At this point we
+ * need to add "d" and all of its parents to the indexableExtrasField. To do so efficiently, the class builds this
+ * branch starting from the leaf and attaches it to parent nodes as we return from each recursive call.
+ */
+public class JsonLogTransformer implements RecordTransformer {
```

Comment:
I wonder how much of the json transformation here is specific to log related application only? Can we refactor the config/transformer so that it can be used for all Json transformation in general? A lot of what this transformer does can be used for non-logging purpose. If possible, I prefer we made a general json transformer.

Reply:
Good point. Instead of calling it `JsonLogTransformer`, we could pick a name that reflects that it takes an input record and transforms it to fit the schema losslessly. Here are some options I can think of:

1. `LosslessSchemaTransformer`
1. `SchemaAdaptationTransformer`
1. `SchemaFitTransformer`
1. `SchemaMoldingTransformer`
1. `TableAdaptationTransformer`

Reply:
No 2 looks like a good one. Should we also add json to the class name and make it JsonSchemaAdaptionTransformer?

Reply:
We could add "Json", but I think this is more general than JSON since it could handle records from any of the decoders like CSV, Parquet, etc. I'm fine with adding "Json" if you think that's best.